File: V5_demo/wss_demo/python/ws_trade_api_demo.py
import asyncio
import websockets
import json
import hmac
import hashlib
from datetime import datetime


async def authenticate(ws):
    api_key = "xxxxxxxxxx"
    api_secret = "xxxxxxxxxx"
    expires = int((datetime.now().timestamp() + 10) * 1000)  # Expires in 10 seconds
    signature = hmac.new(api_secret.encode(), f'GET/realtime{expires}'.encode(), hashlib.sha256).hexdigest()

    auth_msg = {
        "reqId": "auth",
        "op": "auth",
        "args": [api_key, expires, signature]
    }

    await ws.send(json.dumps(auth_msg))
    response = await ws.recv()
    print(f"Auth response: {response}")


async def send_subscription(ws):
    sub_msg = {
        "reqId": "sub-003",
        "header": {
            "X-BAPI-TIMESTAMP": str(int(datetime.now().timestamp() * 1000)),
            "X-BAPI-RECV-WINDOW": "8000",
        },
        "op": "order.create",
        "args": [
            {
                "category": "linear",
                "orderLinkId": "crzyO2.0-1716303565222835550",
                "orderType": "Market",
                "positionIdx": 1,
                "price": "0.5496",
                "qty": "11",
                "side": "Buy",
                "symbol": "XRPUSDT",
                "reduceOnly": False,
                "closeOnTrigger": False
            }
        ]
    }

    await ws.send(json.dumps(sub_msg))
    print("Subscription message sent.")


async def on_ping(ws):
    while True:
        try:
            print('Sending ping')
            await ws.send(json.dumps({'op': 'ping'}))
            await asyncio.sleep(10)  # Sleep for 10 seconds
        except websockets.exceptions.ConnectionClosed:
            print("WebSocket connection closed.")
            break


async def listen_messages(ws):
    try:
        while True:
            message = await ws.recv()
            print(f"Message received: {message}")
    except websockets.exceptions.ConnectionClosed:
        print("WebSocket connection closed.")


async def main():
    uri = "wss://stream-testnet.bybit.com/v5/trade"
    async with websockets.connect(uri) as ws:
        await authenticate(ws)
        await send_subscription(ws)
        listen_task = asyncio.create_task(listen_messages(ws))
        ping_task = asyncio.create_task(on_ping(ws))
        await asyncio.gather(listen_task, ping_task)


# Run the main function in an asyncio event loop
if __name__ == "__main__":
    asyncio.run(main())

File: V5_demo/wss_demo/java/ws_trade_api_demo.java
package com.bybit.api.examples.Websocket.api;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.codec.binary.Hex;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.enums.ReadyState;
import org.java_websocket.handshake.ServerHandshake;
import org.jetbrains.annotations.NotNull;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.URISyntaxException;
import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class ws_trade_api_demo {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static void main(String[] args) {
        try {
            connect();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    static void connect() throws URISyntaxException {
        URI uri = new URI("wss://stream-testnet.bybit.com/v5/trade");
        WebSocketClient ws = getWebSocketClient(uri);
        ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
        executor.scheduleAtFixedRate(() -> {
            try {
                if (ws.getReadyState() == ReadyState.OPEN) {
                    System.out.println("Ping sent");
                    ws.send("{\"op\":\"ping\"}");
                }
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        }, 0, 20, TimeUnit.SECONDS);
    }

    @NotNull
    private static WebSocketClient getWebSocketClient(URI uri) {
        WebSocketClient ws = new WebSocketClient(uri) {
            @Override
            public void onOpen(ServerHandshake handshake) {
                System.out.println("Connected.");
                ws_trade_api_demo.onOpen(this);
            }

            @Override
            public void onMessage(String message) {
                System.out.println("Message received : " + message);
            }

            @Override
            public void onClose(int code, String reason, boolean remote) {
                System.out.println("### about to close ###");
            }

            @Override
            public void onError(Exception e) {
                System.out.println(e.getMessage());
            }
        };

        ws.connect();
        return ws;
    }

    static void onOpen(WebSocketClient ws) {
        System.out.println("opened");
        sendAuth(ws);
        sendSubscription(ws);
    }

    static void sendAuth(WebSocketClient ws) {
        String key = "xxxxxx";
        String secret = "xxxxxxxx";
        long expires = Instant.now().toEpochMilli() + 10000;
        String _val = "GET/realtime" + expires;

        byte[] hmacSha256;
        try {
            SecretKeySpec secretKeySpec = new SecretKeySpec(secret.getBytes(), "HmacSHA256");
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(secretKeySpec);
            hmacSha256 = mac.doFinal(_val.getBytes());
            var args = List.of(key, expires, Hex.encodeHexString(hmacSha256));
            var authMap = Map.of("req_id", UUID.randomUUID().toString(), "op", "auth", "args", args);
            var authJson = objectMapper.writeValueAsString(authMap);
            System.out.println("Auth Message" + authJson);
            ws.send(authJson);
        } catch (Exception e) {
            throw new RuntimeException("Failed to calculate hmac-sha256", e);
        }
    }

    static void sendSubscription(WebSocketClient ws) {
        try {
            Map<String, Object> subMessage = new HashMap<>();
            subMessage.put("reqId", UUID.randomUUID().toString());
            subMessage.put("header", Map.of("X-BAPI-TIMESTAMP", String.valueOf(Instant.now().toEpochMilli()), "X-BAPI-RECV-WINDOW", "8000"));
            subMessage.put("op", "order.create");
            // args is now a List of Map to match the specified format
            List<Map<String, Object>> args = List.of(
                    Map.of(
                            "symbol", "ETHUSDT",
                            "side", "Buy",
                            "orderType", "Limit",
                            "qty", "0.2",
                            "price", "2800",
                            "category", "linear",
                            "timeInForce", "PostOnly"
                    )
            );
            subMessage.put("args", args);
            String json = objectMapper.writeValueAsString(subMessage);
            System.out.println("Trade Message" + json);
            ws.send(json);
        } catch (JsonProcessingException e) {
            System.out.println("Error serializing message: " + e.getMessage());
        }
    }
}
